File: Source/KyndeBlade/Combat/StatusEffect.cpp
#include "Combat/StatusEffect.h"
#include "Characters/MedievalCharacter.h"

UStatusEffect::UStatusEffect()
{
	EffectData = FStatusEffectData();
}

void UStatusEffect::UpdateEffect(float DeltaTime)
{
	if (EffectData.Duration > 0.0f)
	{
		EffectData.RemainingTime -= DeltaTime;
		if (EffectData.RemainingTime <= 0.0f)
		{
			EffectData.RemainingTime = 0.0f;
		}
	}
}

bool UStatusEffect::IsExpired() const
{
	if (EffectData.Duration <= 0.0f)
	{
		return false; // Permanent effect
	}
	return EffectData.RemainingTime <= 0.0f;
}

void UStatusEffect::ApplyEffect(AMedievalCharacter* Target)
{
	if (!Target)
	{
		return;
	}

	ApplyStatModifiers(Target, true);
}

void UStatusEffect::RemoveEffect(AMedievalCharacter* Target)
{
	if (!Target)
	{
		return;
	}

	ApplyStatModifiers(Target, false);
}

void UStatusEffect::StackEffect(int32 AdditionalStacks)
{
	EffectData.StackCount += AdditionalStacks;
	
	// For Hunger, stacking makes it worse
	if (EffectData.EffectType == EStatusEffectType::Hunger)
	{
		// Each stack increases severity
		float StackMultiplier = 1.0f + (EffectData.StackCount - 1) * 0.2f; // 20% worse per stack
		EffectData.AttackPowerModifier = FMath::Max(0.1f, 1.0f - (EffectData.StackCount * 0.15f)); // -15% per stack
		EffectData.DefenseModifier = FMath::Max(0.1f, 1.0f - (EffectData.StackCount * 0.15f));
		EffectData.SpeedModifier = FMath::Max(0.1f, 1.0f - (EffectData.StackCount * 0.15f));
		EffectData.StaminaRegenModifier = FMath::Max(0.0f, 1.0f - (EffectData.StackCount * 0.25f)); // -25% stamina regen per stack
		EffectData.DamagePerSecond = 2.0f * EffectData.StackCount; // 2 DPS per stack
		EffectData.KyndeGenerationModifier = FMath::Max(0.0f, 1.0f - (EffectData.StackCount * 0.3f)); // -30% Kynde generation per stack
	}
}

void UStatusEffect::ApplyStatModifiers(AMedievalCharacter* Target, bool bApply)
{
	if (!Target)
	{
		return;
	}

	// Apply or remove stat modifiers
	// Note: In a full implementation, characters would track base stats separately
	// For now, we apply multipliers directly (effects should be applied in order)
	if (bApply)
	{
		// Apply modifiers
		Target->Stats.AttackPower *= EffectData.AttackPowerModifier;
		Target->Stats.Defense *= EffectData.DefenseModifier;
		Target->Stats.Speed *= EffectData.SpeedModifier;
	}
	else
	{
		// Remove modifiers (divide by the modifier to reverse)
		if (EffectData.AttackPowerModifier > 0.0f)
		{
			Target->Stats.AttackPower /= EffectData.AttackPowerModifier;
		}
		if (EffectData.DefenseModifier > 0.0f)
		{
			Target->Stats.Defense /= EffectData.DefenseModifier;
		}
		if (EffectData.SpeedModifier > 0.0f)
		{
			Target->Stats.Speed /= EffectData.SpeedModifier;
		}
	}
}

void UStatusEffect::ApplyDamageOverTime(AMedievalCharacter* Target, float DeltaTime)
{
	if (!Target || EffectData.DamagePerSecond <= 0.0f)
	{
		return;
	}

	float Damage = EffectData.DamagePerSecond * DeltaTime;
	Target->TakeDamage(Damage, nullptr);
}

// Static factory methods
UStatusEffect* UStatusEffect::CreateHungerEffect(float Duration, int32 Stacks)
{
	UStatusEffect* Effect = NewObject<UStatusEffect>();
	if (Effect)
	{
		Effect->EffectData.EffectType = EStatusEffectType::Hunger;
		Effect->EffectData.Duration = Duration; // 0 = permanent until removed
		Effect->EffectData.RemainingTime = Duration;
		Effect->EffectData.StackCount = Stacks;
		Effect->EffectData.EffectName = TEXT("Hunger");
		
		// Hunger is the WORST status effect - severe penalties
		Effect->EffectData.AttackPowerModifier = FMath::Max(0.1f, 1.0f - (Stacks * 0.15f)); // -15% attack per stack (min 10%)
		Effect->EffectData.DefenseModifier = FMath::Max(0.1f, 1.0f - (Stacks * 0.15f)); // -15% defense per stack (min 10%)
		Effect->EffectData.SpeedModifier = FMath::Max(0.1f, 1.0f - (Stacks * 0.15f)); // -15% speed per stack (min 10%)
		Effect->EffectData.StaminaRegenModifier = FMath::Max(0.0f, 1.0f - (Stacks * 0.25f)); // -25% stamina regen per stack
		Effect->EffectData.DamagePerSecond = 2.0f * Stacks; // 2 DPS per stack
		Effect->EffectData.KyndeGenerationModifier = FMath::Max(0.0f, 1.0f - (Stacks * 0.3f)); // -30% Kynde generation per stack
		Effect->EffectData.StatModifier = FMath::Max(0.1f, 1.0f - (Stacks * 0.2f)); // Overall -20% stats per stack
	}
	return Effect;
}

UStatusEffect* UStatusEffect::CreateFrostEffect(float Duration, float SlowAmount)
{
	UStatusEffect* Effect = NewObject<UStatusEffect>();
	if (Effect)
	{
		Effect->EffectData.EffectType = EStatusEffectType::Frost;
		Effect->EffectData.Duration = Duration;
		Effect->EffectData.RemainingTime = Duration;
		Effect->EffectData.SpeedModifier = SlowAmount;
		Effect->EffectData.EffectName = TEXT("Frost");
	}
	return Effect;
}

UStatusEffect* UStatusEffect::CreateBurningEffect(float Duration, float DamagePerSecond)
{
	UStatusEffect* Effect = NewObject<UStatusEffect>();
	if (Effect)
	{
		Effect->EffectData.EffectType = EStatusEffectType::Burning;
		Effect->EffectData.Duration = Duration;
		Effect->EffectData.RemainingTime = Duration;
		Effect->EffectData.DamagePerSecond = DamagePerSecond;
		Effect->EffectData.EffectName = TEXT("Burning");
	}
	return Effect;
}

UStatusEffect* UStatusEffect::CreatePoisonEffect(float Duration, float DamagePerSecond)
{
	UStatusEffect* Effect = NewObject<UStatusEffect>();
	if (Effect)
	{
		Effect->EffectData.EffectType = EStatusEffectType::Poison;
		Effect->EffectData.Duration = Duration;
		Effect->EffectData.RemainingTime = Duration;
		Effect->EffectData.DamagePerSecond = DamagePerSecond;
		Effect->EffectData.EffectName = TEXT("Poison");
	}
	return Effect;
}

UStatusEffect* UStatusEffect::CreateStunEffect(float Duration)
{
	UStatusEffect* Effect = NewObject<UStatusEffect>();
	if (Effect)
	{
		Effect->EffectData.EffectType = EStatusEffectType::Stun;
		Effect->EffectData.Duration = Duration;
		Effect->EffectData.RemainingTime = Duration;
		Effect->EffectData.SpeedModifier = 0.0f; // Cannot act
		Effect->EffectData.EffectName = TEXT("Stun");
	}
	return Effect;
}

File: Assets/Scripts/Combat/StatusEffect.cs
using System;
using UnityEngine;
using KyndeBlade;

namespace KyndeBlade.Combat
{
    public enum StatusEffectType
    {
        None,
        Hunger,
        Frost,
        Burning,
        Poison,
        Stun,
        Slow,
        Weak,
        Vulnerable,
        Blessed,
        KyndeBoost
    }

    [Serializable]
    public class StatusEffectData
    {
        public StatusEffectType EffectType;
        public float Duration;
        public float RemainingTime;
        public int StackCount = 1;
        public float DamagePerSecond;
        public float AttackPowerModifier = 1f;
        public float DefenseModifier = 1f;
        public float SpeedModifier = 1f;
        public float StaminaRegenModifier = 1f;
        public float KyndeGenerationModifier = 1f;
        public string EffectName;
    }

    public class StatusEffect
    {
        public StatusEffectData Data;

        public StatusEffect()
        {
            Data = new StatusEffectData();
        }

        public void UpdateEffect(float deltaTime)
        {
            if (Data.Duration > 0f)
            {
                Data.RemainingTime -= deltaTime;
                Data.RemainingTime = Mathf.Max(0, Data.RemainingTime);
            }
        }

        public bool IsExpired()
        {
            if (Data.Duration <= 0f) return false;
            return Data.RemainingTime <= 0f;
        }

        public void ApplyEffect(MedievalCharacter target)
        {
            if (target == null) return;
            ApplyStatModifiers(target, true);
        }

        public void RemoveEffect(MedievalCharacter target)
        {
            if (target == null) return;
            ApplyStatModifiers(target, false);
        }

        public void StackEffect(int additionalStacks)
        {
            Data.StackCount += additionalStacks;
            if (Data.EffectType == StatusEffectType.Hunger)
            {
                Data.AttackPowerModifier = Mathf.Max(0.1f, 1f - Data.StackCount * 0.15f);
                Data.DefenseModifier = Mathf.Max(0.1f, 1f - Data.StackCount * 0.15f);
                Data.SpeedModifier = Mathf.Max(0.1f, 1f - Data.StackCount * 0.15f);
                Data.StaminaRegenModifier = Mathf.Max(0f, 1f - Data.StackCount * 0.25f);
                Data.DamagePerSecond = 2f * Data.StackCount;
                Data.KyndeGenerationModifier = Mathf.Max(0f, 1f - Data.StackCount * 0.3f);
            }
        }

        public void ApplyDamageOverTime(MedievalCharacter target, float deltaTime)
        {
            if (target == null || Data.DamagePerSecond <= 0f) return;
            float damage = Data.DamagePerSecond * deltaTime;
            target.ApplyCustomDamage(damage, null);
        }

        void ApplyStatModifiers(MedievalCharacter target, bool apply)
        {
            if (target == null) return;
            if (apply)
            {
                target.Stats.AttackPower *= Data.AttackPowerModifier;
                target.Stats.Defense *= Data.DefenseModifier;
                target.Stats.Speed *= Data.SpeedModifier;
            }
            else
            {
                if (Data.AttackPowerModifier > 0f) target.Stats.AttackPower /= Data.AttackPowerModifier;
                if (Data.DefenseModifier > 0f) target.Stats.Defense /= Data.DefenseModifier;
                if (Data.SpeedModifier > 0f) target.Stats.Speed /= Data.SpeedModifier;
            }
        }

        public static StatusEffect CreateHungerEffect(float duration = 0f, int stacks = 1)
        {
            var effect = new StatusEffect
            {
                Data = new StatusEffectData
                {
                    EffectType = StatusEffectType.Hunger,
                    Duration = duration,
                    RemainingTime = duration,
                    StackCount = stacks,
                    EffectName = "Hunger",
                    AttackPowerModifier = Mathf.Max(0.1f, 1f - stacks * 0.15f),
                    DefenseModifier = Mathf.Max(0.1f, 1f - stacks * 0.15f),
                    SpeedModifier = Mathf.Max(0.1f, 1f - stacks * 0.15f),
                    StaminaRegenModifier = Mathf.Max(0f, 1f - stacks * 0.25f),
                    DamagePerSecond = 2f * stacks,
                    KyndeGenerationModifier = Mathf.Max(0f, 1f - stacks * 0.3f)
                }
            };
            return effect;
        }
    }
}
